src/sonic-host-services/scripts/hostcfgd

```diff
@@ -57,6 +57,157 @@ def run_cmd(cmd, log_err = True):
+
+        # Change auto-restart configuration first.
+        # If service reached failed state before this configuration applies (e.g. on boot)
+        # the next called self.update_feature_state will start it again. If it will fail
+        # again the auto restart will kick-in. Another order may leave it in failed state
+        # and not auto restart.
+        if self._cached_config[feature_name].auto_restart != feature.auto_restart:
+            self.update_feature_auto_restart(feature)
+            self._cached_config[feature_name].auto_restart = feature.auto_restart
+
+        # Enable/disable the container service if the feature state was changed from its previous state.
+        if self._cached_config[feature_name].state != feature.state:
+            self.update_feature_state(feature)
+            self._cached_config[feature_name].state = feature.state
+
+    def update_all_features_config(self):
+        feature_table = self._config_db.get_table('FEATURE')
+        for feature_name in feature_table.keys():
+            if not feature_name:
+                syslog.syslog(syslog.LOG_WARNING, "Feature is None")
+                continue
+
+            self.handle(feature_name, feature_table[feature_name])
+
+    def update_feature_state(self, feature):
+        if self._cached_config[feature.name].state == "always_enabled":
+            syslog.syslog(syslog.LOG_INFO, "Feature '{}' service is always enabled"
+                .format(feature.name))
+            entry = self._config_db.get_entry('FEATURE', feature.name)
+            entry['state'] = 'always_enabled'
+            self._config_db.set_entry('FEATURE', feature.name, entry)
+            return
+
+        # Create feature name suffix depending feature is running in host or namespace or in both
+        feature_name_suffix_list = self._get_feature_name_suffixes(feature)
+
+        if not feature_name_suffix_list:
+            syslog.syslog(syslog.LOG_ERR, "Feature '{}' service not available"
+                          .format(feature.name))
+
+        feature_suffixes = self._get_feature_suffixes(feature)
+
+        if feature.state == "enabled":
+            start_cmds = []
+            for feature_name_suffix in feature_name_suffix_list:
+                for suffix in feature_suffixes:
+                    start_cmds.append("sudo systemctl unmask {}.{}".format(feature_name_suffix, suffix))
+                # If feature has timer associated with it, start/enable corresponding systemd .timer unit
+                # otherwise, start/enable corresponding systemd .service unit
+                start_cmds.append("sudo systemctl enable {}.{}".format(feature_name_suffix, feature_suffixes[-1]))
+                start_cmds.append("sudo systemctl start {}.{}".format(feature_name_suffix, feature_suffixes[-1]))
+                for cmd in start_cmds:
+                    syslog.syslog(syslog.LOG_INFO, "Running cmd: '{}'".format(cmd))
+                    try:
```

Comment:
Can we enhance run_cmd() use it to return error code as well as log the error read from the exception

Reply:
Thanks, yes, please check the enhanced run_cmd()